pkg/performanceprofile/controller/performanceprofile_controller.go

```diff
@@ -449,16 +451,40 @@ func (r *PerformanceProfileReconciler) Reconcile(ctx context.Context, req ctrl.R
 	})
 	err = r.Client.Get(context.Background(), key, nodeCfg)
 	if err != nil {
-		klog.Errorf("failed to get config node object; name=%q err=%v", nodeCfg.GetName(), err)
-		nodeCfg.Name = nodeCfgName
-		nodeCfg.Spec.CgroupMode = apiconfigv1.CgroupModeV1
-		//nolint:errcheck
-		r.Client.Update(ctx, nodeCfg)
+		return reconcile.Result{}, err
 	}
+
 	if nodeCfg.Spec.CgroupMode != apiconfigv1.CgroupModeV1 {
+		klog.Infof("Switching cluster to cgroupv1")
+
+		profileMCP, err := r.getMachineConfigPoolByProfile(ctx, instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		startGeneration := profileMCP.Generation
+
 		nodeCfg.Spec.CgroupMode = apiconfigv1.CgroupModeV1
-		//nolint:errcheck
-		r.Client.Update(ctx, nodeCfg)
+		if err := r.Client.Update(ctx, nodeCfg); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		err = wait.PollUntilContextCancel(context.Background(), 7*time.Second, true, func(ctx context.Context) (done bool, err error) {
+			profileMCP, err := r.getMachineConfigPoolByProfile(ctx, instance)
+			if err != nil {
+				return true, err
+			}
+			klog.Infof("watching for generation change (startGeneration=%v) (mcpGeneration=%v)", startGeneration, profileMCP.Generation)
```

Comment:
Is the generation enough? I can actually still see a race here. The generation is set by the API server iirc, but MCO might not notice that immediately. What we should be watching is generation change + the observedGeneration if the MCP supports that. That way we can be sure MCO already noticed the update.

Reply:
Generation is enough, because the MCO is writing the rendered MachineConfig to the API, and has written to the MCP pool (causing the generation change). This code is watching on the pool generation change. MCO writes an update to the MCP to the API which causes this generation change.